src/lib.rs

```diff
@@ -277,532 +135,100 @@ impl TsunamiBuilder {
-            errors.par_extend(machines.par_iter_mut().flat_map(|(name, machines)| {
-                let f = &setup_fns[name];
-                machines
-                    .par_iter_mut()
-                    .map(move |machine| -> Result<_, Error> {
-                        use std::net::{IpAddr, SocketAddr};
-                        let mut sess = ssh::Session::connect(
-                            &log,
-                            &usernames[name],
-                            SocketAddr::new(
-                                machine
-                                    .public_ip
-                                    .parse::<IpAddr>()
-                                    .context("machine ip is not an ip address")?,
-                                22,
-                            ),
-                            private_key_file.path(),
-                            wait_for,
-                        ).context(format!(
-                            "failed to ssh to {} machine {}",
-                            name, machine.public_dns
-                        ))
-                            .map_err(|e| {
-                                error!(log, "failed to ssh to {}:{}", &name, &machine.public_ip);
-                                e
-                            })?;
+            })
+            .try_reduce(Vec::new, |mut provs, prov| {
+                provs.extend(prov);
+                Ok(provs)
+            })?;
 
-                        debug!(log, "setting up {} instance", name; "ip" => &machine.public_ip);
-                        f(&mut sess)
-                            .context(format!("setup procedure for {} machine failed", name))
-                            .map_err(|e| {
-                                error!(log, "setup for {} machine failed", name);
-                                e
-                            })?;
-                        info!(log, "finished setting up {} instance", name; "ip" => &machine.public_ip);
+        Ok(Tsunami(providers, log))
+    }
+}
 
-                        machine.ssh = Some(sess);
-                        Ok(())
+/// When this is dropped, the instances are all terminated automatically.
+///
+/// # Note
+/// See caveats for Azure and Baremetal machines.
+#[must_use]
+pub struct Tsunami<L: Launcher>(Vec<L>, slog::Logger);
+
+impl<L: Launcher> Tsunami<L> {
+    /// Use to access the machines. The `HashMap` of machines is keyed by the friendly names
+    /// assigned by the call to [add](TsunamiBuilder::add).
+    /// The returned `Machine`s will live for the lifetime of self.
+    pub fn get_machines<'l>(&'l self) -> Result<HashMap<String, Machine<'l>>, Error> {
+        self.0
+            .par_iter()
+            .map(|prov| prov.connect_instances())
+            .try_fold(
+                HashMap::new,
+                |mut machs, res: Result<HashMap<String, Machine<'l>>, Error>| {
+                    res.and_then(|ms| {
+                        machs.extend(ms);
```

Comment:
We should warn somewhere about what happens if a user gives two machines the same name.

Reply:
Or ideally detect this case and error.

Reply:
yep, added a check in `add()` for duplicate keys.